### PR TITLE
Gradient accumulation over multiple batches.

### PR DESCRIPTION
**Pull request by fhieber**
Allows simulation of large batch sizes with gradient updates accumulated from several batches processed in sequence.

Adjusts speed metric loggings and makes sure checkpointing still works w.r.t the chosen interval for number of updates.
It also scales the loss accordingly.

#### Pull Request Checklist ##
- [x] Changes are complete (if posting work-in-progress code, prefix your pull request title with '[WIP]'
until you can check this box.
- [x] Unit tests pass (`pytest`)
- [x] Were system tests modified? If so did you run these at least 5 times to account for the variation across runs?
- [x] System tests pass (`pytest test/system`)
- [x] Passed code style checking (`./style-check.sh`)
- [x] You have considered writing a test
- [x] Updated major/minor version in `sockeye/__init__.py`. Major version bump if this is a backwards incompatible change.
- [x] Updated CHANGELOG.md

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: sockeye/optimizers.py [hunk 1]
```diff
@@ -38,14 +38,16 @@ def __init__(self,
                  kvstore: str,
                  initializer: mx.initializer.Initializer,
                  gradient_clipping_type: str,
-                 gradient_clipping_threshold: Optional[float]) -> None:
+                 gradient_clipping_threshold: Optional[float],
+                 update_interval: Optional[int] = None) -> None:
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/optimizers.py]: right, will change

File: sockeye/train.py [hunk 2]
```diff
@@ -798,10 +800,14 @@ def create_optimizer_config(args: argparse.Namespace, source_vocab_sizes: List[i
...
                              gradient_clipping_type=gradient_clipping_type,
-                             gradient_clipping_threshold=gradient_clipping_threshold)
+                             gradient_clipping_threshold=gradient_clipping_threshold,
+                             update_interval=args.update_interval)
     config.set_lr_scheduler(lr_sched)
     logger.info("Optimizer: %s", config)
     logger.info("Gradient Compression: %s", gradient_compression_params(args))
+    if args.update_interval > 1:
+        logger.info("Gradient accumulation over %d batches. Effective batch size: %d",
+                    args.update_interval, args.batch_size * args.update_interval)
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/train.py]: changed

File: sockeye/training.py [hunk 3]
```diff
@@ -650,12 +661,13 @@ def fit(self,
...
 
-            speedometer(self.state.epoch, self.state.updates, batch_num_samples, batch_num_tokens, metric_train)
+            speedometer(self.state.epoch, self.state.batches, self.state.updates,
+                        batch_num_samples, batch_num_tokens, metric_train)
 
             ############
             # CHECKPOINT
             ############
-            if self.state.updates > 0 and self.state.updates % checkpoint_interval == 0:
+            if self.state.updates > 0 and self.state.batches % (checkpoint_interval * self.update_interval) == 0:
```
[the remark this answers is not in the document]
Comment (fhieber) [reply sockeye/training.py]: This check is executed with each batch, so we need a logic that does not return True `update_interval`-times when `self.state.updates` is unchanged.

[the remark this answers is not in the document]
Comment (tdomhan) [reply sockeye/training.py]: ahh got it. makes sense!
